### PR TITLE
Increase maxIdleConnection limit when creating etcd client in apiserver.

### LINKED ISSUE #7160: Too much time spent in net.(*netFD).dial
When evaluating results of #6877 I have noticed a very strange behavior of apiserver when running density tests on a 100 node cluster with 30 pods per node. It seems that sometimes we spent more than 40% of CPU time in net.(*netFD).dial function.

My suspicion is that it's because after each watch message we terminate old connection to etcd and create a new one. With multiple watchers and many update this can be very expensive.

@davidopp @lavalamp @wojtek-t @xiang90

### PR DESCRIPTION
**Pull request by wojtek-t**
Fix #7160 

cc @fgrzadkowski @lavalamp @xiang90

### CODE REVIEW

**Comment by fgrzadkowski** [comment]
lgtm.

**Comment by fgrzadkowski** [comment]
Will merge on green

**Comment by xiang90** [comment]
lgtm

File: cmd/kube-apiserver/app/server.go [hunk 1]
```diff
@@ -181,6 +182,29 @@ func (s *APIServer) verifyPortalFlags() {
 	}
 }
 
+// dial attempts to open a TCP connection to the provided address, explicitly
+// enabling keep-alives with a one-second interval.
+//
+// TODO: This is copy-pasted from etcd.Client.dial(), since that method isn't public.
```
Comment (lavalamp) [inline cmd/kube-apiserver/app/server.go]: :( Unfortunately, that means it needs to be located in a file in third_party

**Comment by lavalamp** [comment]
I really wanted to merge this but we should probably keep the lawyercats happy.

File: cmd/kube-apiserver/app/server.go (hunk 1 above)
Comment (wojtek-t) [inline cmd/kube-apiserver/app/server.go]: Where exactly I should move it?

Comment (lavalamp) [inline cmd/kube-apiserver/app/server.go]: make a new directory/file in ```third_party/forked/<name of etcd client library>/<name of etcd client library file>.go`` and make it a public function there. Also include the third party LICENSE and any header at the top of the file.

Comment (wojtek-t) [inline cmd/kube-apiserver/app/server.go]: Sorry for my ignorance here - but what should LICENCE file contain?
Any of files under Godeps/_workspace/src/github.com/coreos/go-etcd/etcd doesn't contain any licence-related header.

Comment (lavalamp) [inline cmd/kube-apiserver/app/server.go]: In that case, don't put the license or the header in. :)

Comment (lavalamp) [inline cmd/kube-apiserver/app/server.go]: @sillsm to verify that I'm recalling the instructions correctly...

Comment (wojtek-t) [inline cmd/kube-apiserver/app/server.go]: Done.

**Comment by wojtek-t** [comment]
PTAL

**Comment by wojtek-t** [comment]
@lavalamp I hope that the current directory structure is fine - if not please let me know.

**Comment by lavalamp** [comment]
LGTM

**Comment by smarterclayton** [comment]
Rather than forking this, it might be easier to reuse our existing TLS transport options:

```
	transport, err := client.TransportFor(&client.Config{
		TLSClientConfig: client.TLSClientConfig{
			CertFile: etcdClientInfo.ClientCert.CertFile,
			KeyFile:  etcdClientInfo.ClientCert.KeyFile,
			CAFile:   etcdClientInfo.CA,
		},
		WrapTransport: func(rt http.RoundTripper) http.RoundTripper {
			transport := rt.(*http.Transport)
			dialer := &net.Dialer{
				Timeout: 30 * time.Second,
				// Lower the keep alive for connections.
				KeepAlive: 1 * time.Second,
			}
			// Because watches are very bursty, defends against long delays
			// in watch reconnections.
			transport.MaxIdleConnsPerHost = 500
			transport.Dial = dialer.Dial
			return transport
		},
	})
	etcdClient := etcdclient.NewClient(etcdClientInfo.URLs)
	etcdClient.SetTransport(transport.(*http.Transport))
```

**Pull request merged**